examples/move/crypto/elgamal/sources/example.move

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Example of proving plaintext equivalence of two ElGamal ciphertexts.
+module elgamal::example {
+    use sui::bls12381::{Self, Scalar, G1};
+    use sui::hash::blake2b256;
+    use sui::group_ops::{bytes, equal, Element};
+
+    /// An encryption of group element m under pk is (r*G, r*pk + m) for random r.
+    public struct ElGamalEncryption has drop, store {
+        ephemeral: Element<G1>,
+        ciphertext: Element<G1>,
+    }
+
+    public fun elgamal_decrypt(sk: &Element<Scalar>, enc: &ElGamalEncryption): Element<G1> {
+        let pk_r = bls12381::g1_mul(sk, &enc.ephemeral);
+        bls12381::g1_sub(&enc.ciphertext, &pk_r)
+    }
+
+    /// Basic sigma protocol for proving equality of two ElGamal encryptions.
+    // See https://crypto.stackexchange.com/questions/30010/is-there-a-way-to-prove-equality-of-plaintext-that-was-encrypted-using-different
+    public struct EqualityProof has drop, store {
+        a1: Element<G1>,
+        a2: Element<G1>,
+        a3: Element<G1>,
+        z1: Element<Scalar>,
+        z2: Element<Scalar>,
+    }
+
+    fun fiat_shamir_challenge(
+        pk1: &Element<G1>,
+        pk2: &Element<G1>,
+        enc1: &ElGamalEncryption,
+        enc2: &ElGamalEncryption,
+        a1: &Element<G1>,
+        a2: &Element<G1>,
+        a3: &Element<G1>,
+    ): Element<Scalar> {
+        let mut to_hash = vector::empty<u8>();
+        to_hash.append(*bytes(pk1));
+        to_hash.append(*bytes(pk2));
+        to_hash.append(*bytes(&enc1.ephemeral));
+        to_hash.append(*bytes(&enc1.ciphertext));
+        to_hash.append(*bytes(&enc2.ephemeral));
+        to_hash.append(*bytes(&enc2.ciphertext));
+        to_hash.append(*bytes(a1));
+        to_hash.append(*bytes(a2));
+        to_hash.append(*bytes(a3));
+        let mut hash = blake2b256(&to_hash);
+        // Make sure we are in the right field. Note that for security we only need the lower 128 bits.
+        *vector::borrow_mut(&mut hash, 0) = 0;
```

Comment:
I'm guessing that this doesn't work yet, right?

```suggestion
        hash[0] = 0;
```

Reply:
yes, left-hand positions not supported yet